packages/playwright/src/index.ts

```diff
@@ -45,23 +44,99 @@ if ((process as any)['__pw_initiator__']) {
-  playwright: [async ({}, use) => {
-    await use(require('playwright-core'));
+  _playwrightImpl: [({}, use) => use(require('playwright-core')), { scope: 'worker' }],
+
+  playwright: [async ({ _playwrightImpl }, use) => {
+    const instrumentation: TestLifecycleInstrumentation = {
+      onTestBegin: async () => {
+        artifactsRecorder = new ArtifactsRecorder(_playwrightImpl, tracing().artifactsDir());
+        await artifactsRecorder.willStartTest(currentTestInfo() as TestInfoImpl);
+      },
+      onTestFunctionEnd: async () => {
+        await artifactsRecorder?.didFinishTestFunction();
+      },
+      onTestEnd: async () => {
+        await artifactsRecorder?.didFinishTest();
+        artifactsRecorder = undefined;
+      },
+    };
+    setTestLifecycleInstrumentation(instrumentation);
+
+    const csiListener: ClientInstrumentationListener = {
+      onApiCallBegin: (apiName: string, params: Record<string, any>, frames: StackFrame[], userData: any, out: { stepId?: string }) => {
+        const testInfo = currentTestInfo();
+        if (!testInfo || apiName.includes('setTestIdAttribute'))
+          return { userObject: null };
+        const step = testInfo._addStep({
+          location: frames[0] as any,
+          category: 'pw:api',
+          title: renderApiCall(apiName, params),
+          apiName,
+          params,
+        });
+        userData.userObject = step;
+        out.stepId = step.stepId;
+      },
+      onApiCallEnd: (userData: any, error?: Error) => {
+        const step = userData.userObject;
+        step?.complete({ error });
+      },
+      onWillPause: () => {
+        currentTestInfo()?.setTimeout(0);
+      },
+      onDidCreateBrowserContext: async (context: BrowserContext) => {
+        await artifactsRecorder?.didCreateBrowserContext(context);
+        const testInfo = currentTestInfo();
+        if (testInfo)
+          attachConnectedHeaderIfNeeded(testInfo, context.browser());
+      },
+      onDidCreateRequestContext: async (context: APIRequestContext) => {
+        await artifactsRecorder?.didCreateRequestContext(context);
+      },
+      onWillCloseBrowserContext: async (context: BrowserContext) => {
+        await artifactsRecorder?.willCloseBrowserContext(context);
+      },
+      onWillCloseRequestContext: async (context: APIRequestContext) => {
+        await artifactsRecorder?.willCloseRequestContext(context);
+      },
+    };
+
+    const clientInstrumentation = (_playwrightImpl as any)._instrumentation as ClientInstrumentation;
+    clientInstrumentation.addListener(csiListener);
+
+    if (currentTestInfo()) {
+      // Setup of this fixture happens after the first test starts.
+      await instrumentation.onTestBegin?.();
+    }
+
+    await use(_playwrightImpl);
+
+    if (currentTestInfo()) {
+      // Teardown of this fixture might happen right before the failing test ends
+      // as a part of worker cleanup.
```

Comment:
I'd rephrase it to something along the lines: "If the test fails, worker teardown will destroy this fixture before `onTestEnd();` is called, so we call it here"